inc/ustring.bi

```diff
@@ -0,0 +1,681 @@
+    m_pBuffer[m_BufferLen + 3] = 0
+  end if
+
+  RETURN TRUE
+
+END FUNCTION
+
+
+'***********************************************************************************************
+' The string parameter is appended to the string held in the class. If the internal string
+' buffer overflows, the class will automatically extend it to an appropriate size.
+'***********************************************************************************************
+
+
+PRIVATE SUB DWSTR.Add (BYREF cws AS DWSTR)
+  ''incoming string is already in wide format, simply copy it to the buffer.
+  DIM nLenString AS ulong = cws.m_BufferLen           'lenght in bytes
+  IF nLenString = 0 THEN RETURN
+
+  this.AppendBuffer(cast(ANY PTR, cws), nLenString)   'copy the string into the buffer and update the length
+
+END SUB
+
+PRIVATE SUB DWSTR.Add (BYVAL pwszStr AS WSTRING PTR)
+
+  IF pwszStr = 0 THEN RETURN
+  ''incoming string is already in wide format
+  DIM nLenString AS ulong = .LEN(*pwszStr)            'lenght in characters
+  IF nLenString = 0 THEN RETURN
+
+  this.AppendBuffer(cast(ANY PTR, pwszStr), nLenString * u_size)          'copy the string into the buffer and update the length
+
+END SUB
+
+
+PRIVATE SUB DWSTR.Add (BYREF ansiStr AS STRING, BYVAL nCodePage AS ulong = 0)
+  ''create the wide string from the incoming ansi string
+  IF LEN(ansiStr) = 0 THEN RETURN
+
+  DIM dwLen AS ulong, pbuffer AS wstring PTR
+
+  dwlen = len(ansistr) * 5                            'enough even, if each byte converts to a surrogate pair
+  pbuffer = Allocate(dwLen)
+
+  *pbuffer = ansistr                                  'let FB�s intrinsic conversion do the job
+  dwlen = len(*pbuffer) * u_size
+
+  IF pbuffer THEN
+
+    this.AppendBuffer(pbuffer, dwLen)                 'copy the string into the buffer and update the length
+    Deallocate(pbuffer)
+  END IF
+
+END SUB
+
+
+'***********************************************************************************************
+' All data in the class object is erased. Actually, we only set the buffer length to zero,
+' indicating no string in the buffer. The allocated memory for the buffer is deallocated
+' when the class is destroyed.
+'***********************************************************************************************
+
+
+PRIVATE SUB DWSTR.Clear
+  m_BufferLen = 0
+
+  m_pBuffer[m_BufferLen] = 0                          'mark the end of the string with a double null
+  m_pBuffer[m_BufferLen + 1] = 0
+
+  if u_size > 2 then
+    m_pBuffer[m_BufferLen + 2] = 0
+    m_pBuffer[m_BufferLen + 3] = 0
+  end if
+
+END SUB
+
+
+'***********************************************************************************************
+' Properties
+'***********************************************************************************************
+
+
+PRIVATE PROPERTY DWSTR.Size() AS LONG
+  PROPERTY = u_size
+END PROPERTY
+
+PRIVATE PROPERTY DWSTR.Size (BYVAL nChars AS LONG)
+  IF nChars > -1 THEN u_size = nChars ELSE u_size = -1
```

Comment:
When would `u_size` ever need to change during runtime?  And if it did for some reason, zero (0), probably shouldn't be allowed.  Doesn't that break all the `\ u_size` math in length calculations?